cosmos/x/evm/plugins/txpool/mempool/mempool_reader.go

```diff
@@ -0,0 +1,182 @@
+	defer etp.mu.RUnlock()
+
+	allNonces := etp.senderIndices
+	queued := make(map[common.Address]coretypes.Transactions)
+
+	for sender, list := range allNonces {
+		// get Eth Address of sender
+		addr := cosmlib.EthAddressFromBech32(sender)
+
+		pendingNonce := int64(-1)
+		contiguous := true
+
+		for elem := list.Front(); elem != nil; elem = elem.Next() {
+			ethTx := evmtypes.GetAsEthTx(utils.MustGetAs[sdk.Tx](elem.Value))
+			if ethTx == nil {
+				continue
+			}
+
+			switch {
+			case contiguous && pendingNonce == -1:
+				// When we see a transaction, mark it as the pending nonce.
+				pendingNonce = int64(ethTx.Nonce())
+			case contiguous && int64(ethTx.Nonce()) == pendingNonce+1:
+				// If we are still contiguous and the nonce is the same as the pending nonce, increment the pending nonce.
+				pendingNonce++
+			case contiguous && int64(ethTx.Nonce()) > pendingNonce+1:
+				// If we are still contiguous and the nonce is greater than the pending nonce, we are no longer contiguous.
+				// Add to the queued list.
+				contiguous = false
+				fallthrough
+			default:
+				// All other transactions in the skip list should be queued.
+				queued[addr] = append(queued[addr], ethTx)
+			}
+		}
+	}
+
+	return queued
+}
+
+// Nonce returns the nonce for the given address from the mempool if the address has sent a tx
+// in the mempool.
+func (etp *EthTxPool) Nonce(addr common.Address) uint64 {
+	etp.mu.RLock()
+	defer etp.mu.RUnlock()
+
+	var pendingNonce int64 = -1
+	// search the addr's txs for the first eth tx nonce (first pending nonce)
+	if txs := etp.senderIndices[cosmlib.AddressToAccAddress(addr).String()]; txs != nil {
+		for elem := txs.Front(); elem != nil; elem = elem.Next() {
+			if ethTx := evmtypes.GetAsEthTx(utils.MustGetAs[sdk.Tx](elem.Value)); ethTx != nil {
+				switch {
+				case pendingNonce == -1:
+					// When we see a transaction, mark it as the pending nonce.
+					pendingNonce = int64(ethTx.Nonce())
+					// If on the first lookup the nonce delta is more than 0, then there is a gap and thus no pending transactions,
+					// but there are queued transactions.
+					if sdbNonce := etp.nr.GetNonce(addr); uint64(pendingNonce)-sdbNonce >= 1 {
+						return sdbNonce
+					}
+				case int64(ethTx.Nonce()) == pendingNonce+1:
+					// If we are still contiguous and the nonce is the same as the pending nonce, increment the pending nonce.
+					pendingNonce++
+				case int64(ethTx.Nonce()) > pendingNonce+1:
+					// As soon as we see a non contiguous nonce we break.
+					return uint64(pendingNonce + 1)
+				}
+			}
+		}
+	}
+
+	// if the addr has no eth txs, fallback to the nonce retriever db
+	if pendingNonce == -1 {
+		pendingNonce = int64(etp.nr.GetNonce(addr))
+	} else {
+		pendingNonce++
+	}
+	return uint64(pendingNonce)
+}
+
+// Stats returns the number of currently pending (locally created) transactions.
+func (etp *EthTxPool) Stats() (int, int) {
+	pending, queued := etp.Content()
+	return len(pending), len(queued)
```

Comment:
`pending` and `queued` are `map[common.Address]coretypes.Transactions`. `len(pending` and `len(queued)` will return # of keys, meaning # of accounts that have sent transactions, not the # of transactions.

Reply:
```
func (etp *EthTxPool) Stats() (int, int) {
        pendingTxsLen := 0
        queuedTxsLen := 0
	pending, queued := etp.Content()
	for _, txs := range pending {
                pendingTxsLen += len(txs)
        }
	for _, txs := range queued {
                queuedTxsLen += len(txs)
        }
	return pendingTxsLen, queuedTxsLen
```

Reply:
good catch